File: packages/types/src/events.ts
// ─── Scan Event ───────────────────────────────────────────────

export type ScanResult = 'valid' | 'invalid' | 'override';
export type BarcodeType = 'GS1-128' | 'Code128' | 'QR' | 'DataMatrix';

export interface ScanEvent {
  id: string;
  tenant_id: string;
  worker_id: string;
  task_id: string | null;
  flow_id: string | null;
  step_id: string | null;
  barcode_raw: string;
  barcode_type: BarcodeType | null;
  parsed_data: Record<string, unknown> | null;
  result: ScanResult;
  rejection_reason: string | null;
  device_id: string | null;
  timestamp: Date;
}

// ─── Realtime Events (SSE) ───────────────────────────────────

export interface TaskAssignedEvent {
  type: 'task_assigned';
  task_id: string;
  worker_id: string;
  flow_id: string;
  tenant_id: string;
}

export interface TaskReassignedEvent {
  type: 'task_reassigned';
  task_id: string;
  from_worker_id: string;
  to_worker_id: string;
  tenant_id: string;
}

export interface SupervisorMessageEvent {
  type: 'supervisor_message';
  worker_id: string;
  message: string;
  tenant_id: string;
}

export interface WaveReleasedEvent {
  type: 'wave_released';
  wave_id: string;
  tenant_id: string;
  task_count: number;
}

export interface StepChangedEvent {
  type: 'step_changed';
  session_id: string;
  worker_id: string;
  step_id: string;
  tenant_id: string;
}

export type RealtimeEvent =
  | TaskAssignedEvent
  | TaskReassignedEvent
  | SupervisorMessageEvent
  | WaveReleasedEvent
  | StepChangedEvent;

File: android/app/src/main/java/com/kinetic/wms/data/model/Events.kt
package com.kinetic.wms.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class BarcodeType {
    @SerialName("GS1-128") GS1_128,
    @SerialName("Code128") CODE_128,
    @SerialName("QR") QR,
    @SerialName("DataMatrix") DATA_MATRIX,
}

@Serializable
data class ScanValidateRequest(
    val barcode: String,
    @SerialName("expected_sku") val expectedSku: String? = null,
    @SerialName("task_id") val taskId: String? = null,
    @SerialName("step_id") val stepId: String? = null,
    @SerialName("flow_id") val flowId: String? = null,
    @SerialName("barcode_type") val barcodeType: BarcodeType? = null,
    @SerialName("device_id") val deviceId: String? = null,
)

@Serializable
data class ScanValidateResponse(
    val sku: String,
    val name: String,
    val uom: String,
    @SerialName("lot_tracked") val lotTracked: Boolean = false,
    @SerialName("serial_tracked") val serialTracked: Boolean = false,
    val lot: String? = null,
)

@Serializable
data class ScanErrorResponse(
    val error: String,
    val barcode: String,
)

@Serializable
sealed class RealtimeEvent {
    @Serializable
    @SerialName("task_assigned")
    data class TaskAssigned(
        @SerialName("task_id") val taskId: String,
        @SerialName("worker_id") val workerId: String,
        @SerialName("flow_id") val flowId: String,
        @SerialName("tenant_id") val tenantId: String,
    ) : RealtimeEvent()

    @Serializable
    @SerialName("task_reassigned")
    data class TaskReassigned(
        @SerialName("task_id") val taskId: String,
        @SerialName("from_worker_id") val fromWorkerId: String,
        @SerialName("to_worker_id") val toWorkerId: String,
        @SerialName("tenant_id") val tenantId: String,
    ) : RealtimeEvent()

    @Serializable
    @SerialName("supervisor_message")
    data class SupervisorMessage(
        val message: String,
        @SerialName("worker_id") val workerId: String,
        @SerialName("tenant_id") val tenantId: String,
    ) : RealtimeEvent()

    @Serializable
    @SerialName("wave_released")
    data class WaveReleased(
        @SerialName("wave_id") val waveId: String,
        @SerialName("task_count") val taskCount: Int,
        @SerialName("tenant_id") val tenantId: String,
    ) : RealtimeEvent()
}
